3p/ampanalytics-lib.js

```diff
@@ -0,0 +1,212 @@
+ * limitations under the License.
+ */
+
+import './polyfills';
+import {tryParseJson} from '../src/json';
+import {dev, initLogConstructor, setReportError} from '../src/log';
+import {IframeMessagingClient} from './iframe-messaging-client';
+import {AMP_ANALYTICS_3P_MESSAGE_TYPE} from '../src/3p-analytics-common';
+
+initLogConstructor();
+// TODO(alanorozco): Refactor src/error.reportError so it does not contain big
+// transitive dependencies and can be included here.
+setReportError(() => {});
+
+/** @private @const {string} */
+const TAG_ = 'ampanalytics-lib';
+
+/**
+ * Receives messages bound for this cross-domain iframe, from all creatives
+ */
+class AmpAnalytics3pMessageRouter {
+
+  /** @param {!Window} win */
+  constructor(win) {
+    /** @private {!Window} */
+    this.win_ = win;
+
+    // The sentinel is required by iframe-messaging-client, and is used to
+    // uniquely identify the frame as part of message routing
+    /** @const {string} */
+    this.sentinel_ = dev().assertString(
+        tryParseJson(this.win_.name, {}).sentinel,
+        'Invalid/missing sentinel on iframe name attribute' + this.win_.name);
+    if (!this.sentinel_) {
+      return;
+    }
+
+    /**
+     * Multiple creatives on a page may wish to use the same type of
+     * amp-analytics tag. This object provides a mapping between the
+     * IDs which identify which amp-analytics tag a message is to/from,
+     * with each ID's corresponding AmpAnalytics3pCreativeMessageRouter,
+     * which is an object that handles messages to/from a particular creative.
+     * @private {!Object<string, !AmpAnalytics3pCreativeMessageRouter>}
+     */
+    this.creativeMessageRouters_ = {};
+
+    /**
+     * Handles communication between frames
+     * @private {!IframeMessagingClient}
+     */
+    this.iframeMessagingClient_ = new IframeMessagingClient(win);
+    this.iframeMessagingClient_.setSentinel(this.sentinel_);
+    this.iframeMessagingClient_.registerCallback(
+        AMP_ANALYTICS_3P_MESSAGE_TYPE.CREATIVE,
+        messageContainer => {
+          let entries;
+          dev().assert(
+              messageContainer.data &&
+          (entries = Object.entries(messageContainer.data)).length,
+              'Received empty new creative message');
+          entries.forEach(entry => {
+            const creativeId = entry[0];
+            const extraData = entry[1];
+            dev().assert(!this.creativeMessageRouters_[creativeId],
+                'Duplicate new creative message for ' + creativeId);
+            this.creativeMessageRouters_[creativeId] =
+            new AmpAnalytics3pCreativeMessageRouter(
+              this.win_, this.iframeMessagingClient_, creativeId, extraData);
+          });
+        });
+    this.iframeMessagingClient_.registerCallback(
+        AMP_ANALYTICS_3P_MESSAGE_TYPE.EVENT,
+        messageContainer => {
+          Object.entries(messageContainer.data).forEach(entry => {
```

Comment:
Done, in the style of lines 70-74 since messageContainer.data is an object, not an array.